chacha20/src/nivc.nr

```diff
@@ -0,0 +1,334 @@
+    state.to_bytes()
+}
+
+// Helper function to check if a byte is padding
+fn is_padding(byte: u8) -> bool {
+    byte == 0xFF
+}
+
+// Helper function to convert little endian to big endian
+fn little_to_big_endian(word: u32) -> u32 {
+    ((word & 0xFF) << 24)
+        | ((word & 0xFF00) << 8)
+        | ((word & 0xFF0000) >> 8)
+        | ((word & 0xFF000000) >> 24)
+}
+
+fn big_to_little_endian(word: u32) -> u32 {
+    ((word & 0xFF) << 24)
+        | ((word & 0xFF00) << 8)
+        | ((word & 0xFF0000) >> 8)
+        | ((word & 0xFF000000) >> 24)
+}
+
+fn get_byte_from_word(word: u32, byte_idx: u8) -> u8 {
+    // ex) word = [00000001, 00000010, 00000011, 00000100]
+    // byte_idx = 0 -> 00000001
+    // byte_idx = 1 -> 00000010
+    // byte_idx = 2 -> 00000011
+    // byte_idx = 3 -> 00000100
+    let byte = (word >> (24 - byte_idx * 8)) & 0xFF;
+    byte as u8
+}
+
+// Structure for NIVC state
+pub struct NivcState<let DATA_BYTES: u32, let PUBLIC_IO_LENGTH: u32> {
+    key: [u32; 8],
+    nonce: [u32; 3],
+    counter: u32,
+    plaintext: [u8; DATA_BYTES],
+    ciphertext_digest: Field,
+}
+
+impl<let DATA_BYTES: u32, let PUBLIC_IO_LENGTH: u32> NivcState<DATA_BYTES, PUBLIC_IO_LENGTH> {
+    fn verify(self, step_in: [Field; PUBLIC_IO_LENGTH]) -> [Field; PUBLIC_IO_LENGTH] {
+        let mut step_out = step_in;
+        // let mut ciphertext_digest_pow = [Field::default(); DATA_BYTES];
+        let mut ciphertext_digest_pow = Field::default();
+        ciphertext_digest_pow = step_in[1];
+
+        // Initialize arrays
+        let mut counter = self.counter;
+        let mut ciphertext: [u32; (DATA_BYTES / 4)] = [0; (DATA_BYTES / 4)];
+        let mut plaintext_bits: [u32; (DATA_BYTES / 4)] = [0; (DATA_BYTES / 4)];
+        let mut is_padding: [bool; DATA_BYTES] = [false; DATA_BYTES];
+        let mut zeroed_plaintext: [u8; DATA_BYTES] = [0; DATA_BYTES];
+        let mut big_endian_ciphertext: [u8; DATA_BYTES] = [0; DATA_BYTES];
+
+        // Convert plaintext to bits with padding handling
+        for i in 0..(DATA_BYTES / 4) {
+            // Process each byte in the word
+            let mut word: [u8; 4] = [0; 4];
+            for j in 0..4 {
+                let byte_idx = i * 4 + j;
+                let is_padding_current = self.plaintext[byte_idx] == 0; // TODO: fix to -1.
+                is_padding[byte_idx] = is_padding_current;
+
+                if is_padding_current {
+                    word[j] = 0;
+                    zeroed_plaintext[byte_idx] = 0;
+                } else {
+                    word[j] = self.plaintext[byte_idx];
+                    zeroed_plaintext[byte_idx] = word[j];
+                }
+            }
+            // concat 4 u8 into 1 u32
+            let concatenated = (word[0] as u32) << 24
+                | (word[1] as u32) << 16
+                | (word[2] as u32) << 8
+                | (word[3] as u32);
+            let le_word = big_to_little_endian(concatenated);
```

Comment:
You can parse in little-endian at first rather than parse in big-endian and convert to little-endian.
```suggestion
            let le_word = (word[3] as u32) << 24
                | (word[2] as u32) << 16
                | (word[1] as u32) << 8
                | (word[0] as u32);
            
            plaintext_bits[i] = le_word;
```

Reply:
or you can use this in Field.
https://noir-lang.org/docs/v1.0.0-beta.2/noir/concepts/data_types/fields#to_le_bytes
Let's research how to count the number of constraints and compare.
https://aztec.network/blog/unconstrained-functions-in-noir

Reply:
``` noir
  let concatenated = (word[0] as u32) << 24
                | (word[1] as u32) << 16
                | (word[2] as u32) << 8
                | (word[3] as u32);
```
1. This is nice!

2. 'to_le_bytes' has a return type of Field.
The type we need to get is u32, so it's hard to use to_le_bytes directly.